wordpress.org/public_html/wp-content/plugins/plugin-directory/class-plugin-search.php

```diff
@@ -0,0 +1,404 @@
+		$export = var_export($expression, TRUE);
+		$patterns = [
+			"/array \(/" => '[',
+			"/^([ ]*)\)(,?)$/m" => '$1]$2',
+			"/=>[ ]?\n[ ]+\[/" => '=> [',
+			"/([ ]*)(\'[^\']+\') => ([\[\'])/" => '$1$2 => $3',
+		];
+		$export = preg_replace(array_keys($patterns), array_values($patterns), $export);
+		if ((bool)$return) return $export; else echo $export;
+	}
+
+	public function option_jetpack_active_modules( $modules ) {
+		if ( self::USE_OLD_SEARCH ) {
+			if ( $i = array_search( 'search', $modules ) )
+				unset( $modules[$i] );
+		} else {
+			$modules[] = 'search';
+		}
+
+		return array_unique( $modules );
+	}
+
+	public function option_has_jetpack_search_product( $option ) {
+		if ( !self::USE_OLD_SEARCH ) {
+			return true;
+		}
+		return $option;
+	}
+
+	/* Make sure the search module is available regardless of Jetpack plan.
+	 * This works because search indexes were manually created for w.org.
+	 */
+	public function jetpack_get_module( $module, $slug ) {
+		if ( !self::USE_OLD_SEARCH ) {
+			if ( 'search' === $slug && isset( $module[ 'plan_classes' ] ) && !in_array( 'free', $module[ 'plan_classes' ] ) ) {
+				$module[ 'plan_classes' ][] = 'free';
+			}
+		}
+
+		return $module;
+	}
+
+	public function jetpack_search_es_wp_query_args( $args, $query ) {
+
+		// Block Search.
+		$this->is_block_search = !empty( $query->query['block_search'] );
+		if ( $this->is_block_search ) {
+			$args['block_search'] = $query->query['block_search'];
+		}
+
+		// How much weighting to put on the Description field.
+		// Blocks get a much lower value here, as it's more title/excerpt (short description) based.
+		$this->desc_boost = $this->is_block_search ? 0.05 : 1;
+
+		// Because most plugins don't have any translations we need to
+		// correct for the very low scores that locale-specific fields.
+		// end up getting. This is caused by the average field length being
+		// very close to zero and thus the BM25 alg discounts fields that are
+		// significantly longer.
+		//
+		// As of 2017-01-23 it looked like we were off by about 10,000x,
+		// so rather than 0.1 we use a much smaller multiplier of en content
+		$this->en_boost             = 0.00001;
+		$this->desc_en_boost        = $this->desc_boost * $this->en_boost;
+
+		// We need to be locale aware for this
+		$this->locale = strtolower( substr( get_locale(), 0, 2 ) );
+		$this->is_english = ( !$this->locale || 'en' === $this->locale );
+
+		if ( $this->is_english ) {
+			$matching_fields      = array(
+				'all_content_en',
+			);
+		} else {
+			$matching_fields      = array(
+				'all_content_' . $this->locale,
+				'all_content_en^' . $this->en_boost,
+			);
+		}
+
+		$args['query_fields'] = $matching_fields;
+
+
+
+
+		return $args;
+	}
+
+	public function jetpack_search_es_query_args( $es_query_args, $query ) {
+		// These are the things that jetpack_search_es_wp_query_args doesn't let us change, so we need to filter the es_query_args late in the code path to add more custom stuff.
```

Comment:
What about just bypassing everything that jetpack_search_es_wp_query_args does and building the whole query in here (completely ignore what jetpack_search_es_wp_query_args has given us).

I think the query could be much easier to review then and the current code feels very fragile in case anything above it changes.

I think you should use the query builder class to create the actual query. 

So effectively you fork the code from here https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/modules/search/class.jetpack-search.php#L912 and then use that to replace the ES query via this filter. The parser/builder takes care of the syntax for you and makes a lot of it more robust to changes.